managed_bytes, managed_bytes_view: implement interface

bhalevy opened the pull request:
Based on top of https://github.com/scylladb/scylla/pull/7490

Implement various interfaces unimplemented in the skeleton implementation.

Commit:
cql3/expr: Use Boost concept assert

In bd6855e, we reverted to Boost ranges and commented out the concept
check.  But Boost has its own concept check, which this patch enables.

Tests: unit (dev)

Signed-off-by: Dejan Mircevski <dejan@scylladb.com>

Closes #7471

bhalevy committed:
test: imr_test should run automatically

Unclear why it was places in test/manual in
commit 1c8736f99823a63bbde104c24ad2ea666290d56e

Test: boost/imr_test

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>
Message-Id: <20201022093826.12009-1-bhalevy@scylladb.com>

avikivity committed:
cql3: expression: mark do_get_value() as static

It is used only later in this file.

avikivity committed:
cql3: expression: linearize get_value_from_mutation() eariler

do_get_value() is careful to return a fragmented view, but
its only caller get_value_from_mutation() linearizes it immediately
afterwards. Linearize it sooner; this prevents mixing in
fragmented values from cells (now via IMR) and fragmented values
from partition/clustering keys. It only works now because
keys are not fragmented outside LSA, and value_view has a special
case for single-fragment values.

This helps when keys become fragmented.

avikivity committed:
bytes: add to_bytes(bytes)

Converting from bytes to bytes is nonsensical, but it helps
when transitioning to other types (managed_bytes/managed_bytes_view),
and these types will have to_bytes() conversions.

avikivity committed:
clustering_bounds_comparator: do not depend on implicit conversion of keys to bytes_view

This implicit conversion will soon be gone, as it is dangerous.
Ask for the representation explicitly.

avikivity committed:
thrift: do not depend on implicit conversion of keys to bytes_view

This implicit conversion will soon be gone, as it is dangerous.
Ask for the representation explicitly.

avikivity committed:
utils: managed_bytes: extract do_linearize_pure()

do_linearize() is an impure function as it changes state
in linearization_context. Extract the pure parts into a new
do_linearize_pure(). This will be used to linearize managed_bytes
without a linearization_context, during the transition period where
fragmented and non-fragmented values coexist.

avikivity committed:
utils: managed_bytes: avoid internal uses of managed_bytes::data()

We use managed_bytes::data() in a few places when we know the
data is non-fragmented (such as when the small buffer optimization
is in use). We'd like to remove managed_bytes::data() as linearization
is bad, so in preparation for that, replace internal uses of data()
with the equivalent direct access.
won't be, so stop using it.

avikivity committed:
utils: managed_bytes: constrain with_linearized_managed_bytes()

The passed function must be called with a no parameters; document
and enforce it.

avikivity committed:
utils: managed_bytes: introduce with_linearized()

This is a temporary scaffold for weaning ourselves off
linearization. It differs from with_linearized_managed_bytes in
that it does not rely on the environment (linearization_context)
and so is easier to remove.

avikivity committed:
test: logalloc_test: don't cast managed_bytes to bytes_view

This linearizing conversion will be removed, so don't use it.
Compare as managed_bytes_view instead.

avikivity committed:
test: sstable_mutation_test: don't rely on managed_bytes->bytes_view conversion

The code relies on conversion chain that relies on managed_bytes
decaying to bytes_view, but this is going away soon. Avoid that chain.

avikivity committed:
types: add to_managed_bytes(const sstring&)

This is a helper for tests (similar to to_bytes(const sstring&).

avikivity committed:
types: change equal() to accept managed_bytes_view

bytes_view can convert to managed_bytes_view, so the change
is compatible with the existing representation and the next
patch, which changes compount types to use managed_bytes_view.

avikivity committed:
utils: managed_bytes: introduce managed_bytes_view

managed_bytes_view is a non-owning view into managed_bytes,
but can also be converted from bytes_view and bytes.

Since it is a bytes_view replacement, it contains similar
methods (substr() and remove_prefix(), as used in the code
base) and a formatting operator. There is a maze of conversions
and comparison operators to avoid ambiguities when the compiler
needs to choose a path to arrive at the correct types.

avikivity committed:
serializer: avoid implicit linearization of managed_bytes

Deserialization and serialization now implicily (or semi
expicitly) linearize by calling managed_bytes::begin()
and managed_bytes::operator bytes_view. Make this explicit
by calling the constructor and managed_bytes::with_linearized().
This makes is easy to remove implicit linearization, and we can
fix the explicit linearizations later.

avikivity committed:
types: implement read_simple*() variants for managed_bytes_view

These follow the bytes_view variants but are implemented for
managed_bytes_view instead. They are needed for deserializing
from managed_bytes_view.

bhalevy committed:
managed_bytes: implement constructor taking managed_bytes_view

Use managed_bytes_view::remove_prefix to iterate through it.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

avikivity committed:
types: abstract_type: add parallel interface with managed_bytes_view instead of bytes_view

For every method that takes a bytes_view, add a corresponding method
that takes a managed_bytes_view, so we can stop requiring that the
caller linearizes.

For now, the new implementations call the old ones, so we don't avoid any
linearizing, but we can later replace those implementations with
non-linearizing ones.

Once the transition is complete, we'll be able to drop the bytes_view
overloads.

avikivity committed:
keys, compound: switch from bytes_view to managed_bytes_view

The keys classes (partition_key et al) already use managed_bytes,
but they assume the data is not fragmented and make liberal use
of that by casting to bytes_view. The view classes use bytes_view.

Change that to managed_bytes_view, and adjust return values
to managed_bytes/managed_bytes_view. To ease the transition
new overloads are added rather than the methods changed.

The callers are adjusted. In some places linearization (to_bytes())
is needed, but this isn't too bad as keys are always <= 64k and thus
will not be fragmented when out of LSA. We can remove this
linearization later.

The serialize_value() template is called from a long chain, and
can be reached with either bytes_view or managed_bytes_view.
Rather than trace and adjust all the callers, we patch it now
with constexpr if.

operator bytes_view (in keys) is converted to operator
managed_bytes_view, allowing callers to defer or avoid
linearization.

bhalevy committed:
managed_bytes: implement managed_bytes_view with_linearized and to_bytes

Define do_linearize_pure that linearizes the managed_bytes_view into
a given buffer.

Use it for with_linearized that returns an optimized bytes_view either to
managed_bytes_view::_current_fragment, if there are no _next_fragments
or to a linearized buffer.

do_linearize_pure is also used for to_bytes to linearize the managed_bytes_view
into a newly allocated `bytes` buffer.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
managed_bytes_view: implement comperators

Selectively expose managed_bytes_view_fragment_iterator::remove_prefix
inherited from managed_bytes_view_base to class managed_bytes_view
and compare_unsigned().

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

avikivity committed:
types: add constraint on lexicographical_try_compare()

Verify that the input types are iterators and their value
types are compatible with the compare function.

avikivity committed:
utils: managed_bytes: remove linearizing accessors

Accessor that require linearization, such as data(), begin(),
and casting to bytes_view, are no longer used and are now removed.

bhalevy committed:
test: add managed_bytes_test

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

Commit:
memtable, row_cache: remove `with_linearized_managed_bytes` uses

Since `managed_bytes::data()` is deleted as well as other public
APIs of `managed_bytes` which would linearize stored values except
for explicit `with_linearized`, there is no point
invoking `with_linearized_managed_bytes` hack which would trigger
automatic linearization under the hood of managed_bytes.

Remove useless `with_linearized_managed_bytes` wrapper from
memtable and row_cache code.

Signed-off-by: Pavel Solodovnikov <pa.solodovnikov@scylladb.com>

bhalevy committed:
managed_bytes_view: implement remove_prefix

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
managed_bytes_view: mark constructors noexcept

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
test: managed_bytes_view: test equality type matrix

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

Commit:
treewide: remove remaining `with_linearized_managed_bytes` uses

There is no point in calling the wrapper since linearization code
is private in `managed_bytes` class and there is no one to call
`managed_bytes::data` because it was deleted recently.

This patch is a prerequisite for removing
`with_linearized_managed_bytes` function completely, alongside with
the corresponding parts of implementation in `managed_bytes`.

Signed-off-by: Pavel Solodovnikov <pa.solodovnikov@scylladb.com>

Commit:
misc: fix indentation

The patch fixes indentation issues introduced in previous patches
related to removing `with_linearized_managed_bytes` uses from the
code tree.

Signed-off-by: Pavel Solodovnikov <pa.solodovnikov@scylladb.com>

Commit:
utils: remove unused linearization facilities in `managed_bytes` class

Remove the following bits of `managed_bytes` since they are unused:
* `with_linearized_managed_bytes` function template
* `linearization_context_guard` RAII wrapper class for managing
  `linearization_context` instances.
* `do_linearize` function
* `linearization_context` class

Since there is no more public or private methods in `managed_class`
to linearize the value except for explicit `with_linearized()`,
which doesn't use any of aforementioned parts, we can safely remove
these.

Signed-off-by: Pavel Solodovnikov <pa.solodovnikov@scylladb.com>

bhalevy committed:
test: managed_bytes_test: test more managed_bytes_view constructors

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
managed_bytes_view: implement substr

Note that _size may be truncated by substr()
so we may need to trim _current_fragment
and drop _next_fragments.

managed_bytes_view: substr: default len to npos

Means return a substr starting from offset until the
end of the managed_bytes_view (same as remove_prefix)

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
managed_bytes_view: implement operator<<

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
managed_bytes_view: add fragment_iterator

Derived from managed_bytes_view_base and uses
remove_prefix to iterate through the fragments.

The reason we use the rather complicated `remove_prefix`
methods is to prepare for `substr` that may truncate the
overall _size, so the last fragment may need truncating
during iteration.

Note that operator++() is marked as noexcept even though
remove_prefix may theoretically throw, since
it must not throw in this path and doing so
indicates an internal error.

Define managed_bytes_fragment_range_view
conforming to the FragmentRange concept and add
respective unit test.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
managed_bytes_view: implement appending_hash

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
managed_bytes: implement hash

For now, implement only hash for managed_bytes_view
and convert managed_bytes to managed_bytes_view
to calculate its hash.

Linearize the bytes first for now, for simplicity
and copleteness, despite being ineficient for
multiple segments.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
managed_bytes_view: split out managed_bytes_view_base

To be used for implementing fragment_iterator.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

Commit:
row_cache: more indentation fixes

Fixup indentation issues introduced in recent patches.

Signed-off-by: Pavel Solodovnikov <pa.solodovnikov@scylladb.com>

avikivity commented on test/boost/managed_bytes_test.cc:
```diff
@@ -325,3 +326,29 @@ SEASTAR_THREAD_TEST_CASE(test_managed_bytes_view_substr) {
+    appending_hash<bytes>{}(h1, b);
+    appending_hash<managed_bytes_view>{}(h2, mv);
+    BOOST_REQUIRE_EQUAL(h1.finalize(), h2.finalize());
+}
```
What does this test? In all cases the managed_bytes will not be fragmented.

bhalevy replied:
Correct, I'll change to convert via `managed_bytes`.
Is there a way to enforce  fragmentation of `managed_bytes`? (command line option maybe?)

avikivity replied:
No. I suggest adding `static managed_bytes managed_bytes::with_fragment_sizes(std::vector<size_t>)` for tests.

avikivity replied:
I'll add auto-fragmentation during normal operation, but the API above can still be useful for tests.

bhalevy replied:
@avikivity maybe just support passing an optional allocation function to the constructor.
Then in the test we can use it to control the allocated fragment sizes whichever way we like.

avikivity commented on utils/managed_bytes.hh:
```diff
@@ -356,6 +357,14 @@ public:
+    using fragment_iterator = managed_bytes_view_fragment_iterator;
+    using const_iterator = fragment_iterator;
+    const_iterator begin() const noexcept;
+    const_iterator end() const noexcept;
```
This is wrong, users of managed_bytes will expect that iterators iterate over bytes, not fragments, since managed_bytes is a bytes replacement.

bhalevy replied:
The fragment_type is `bytes_view`, isn't that expected when iterating over `bytes`?

avikivity replied:
When iterating over bytes, you get a byte, not a byte_view.

denesb replied:
I like the adaptor idea, we can create a friend class that does the adaptation using the managed_bytes internals directly.

avikivity replied:
Yes, and managed_bytes[_view]::as_fragment_range() can return it.

denesb replied:
I don't think we can have it as static member. The adaptor has to own the `managed_bytes{_view}` instance it operates on. Adaptors are expected to own the range they adapt. One of the bugs I fixed in IMR was caused by some owning the range some not.

bhalevy replied:
do we also need the byte iterator?

bhalevy replied:
> I don't think we can have it as static member. The adaptor has to own the `managed_bytes{_view}` instance it operates on. Adaptors are expected to own the range they adapt. One of the bugs I fixed in IMR was caused by some owning the range some not.

I implemented it as returning a `managed_bytes_fragment_range_view(*this)` that holds a copy of the `managed_bytes_view`.

denesb replied:
A copy works for the view, but we also need support for the `managed_bytes` itself.

denesb replied:
> do we also need the byte iterator?

I don't think so.

denesb replied:
Or maybe we do, I just came across some code that would be much better with it.

avikivity commented on utils/managed_bytes.hh:
```diff
@@ -356,6 +357,14 @@ public:
+    using const_iterator = fragment_iterator;
+    const_iterator begin() const noexcept;
+    const_iterator end() const noexcept;
+    size_t size_bytes() const noexcept { return size(); }
```
managed_bytes::size() returns the number of bytes, not the number of fragments. It's wrong to make managed_bytes a FragmentRange, the users expect it to behave as a container of bytes rather than a container of fragments. Need an adapter class for that.

avikivity commented on utils/managed_bytes.hh:
```diff
@@ -479,36 +479,36 @@ std::invoke_result_t<Func, bytes_view> managed_bytes_view::with_linearized(Func&
 // These are used to resolve ambiguities because managed_bytes_view can be converted to managed_bytes and vice versa
-inline bool operator==(const managed_bytes& a, const managed_bytes_view_base& b) {
+inline bool operator==(const managed_bytes& a, const managed_bytes_view_base& b) noexcept {
     return managed_bytes_view(a) == b;
```
Let's leave these until later to reduce noise. They don't really add anything.

bhalevy replied:
ok

avikivity commented on test/boost/managed_bytes_test.cc:
```diff
@@ -122,3 +122,29 @@ SEASTAR_THREAD_TEST_CASE(test_managed_bytes_view_from_bytes) {
+    b = random_bytes(1, 16);
+    m = managed_bytes_view(b);
+    n = tests::random::get_int(size_t(0), b.size());
+    BOOST_TEST_MESSAGE(format("size={} remove_prefix={}", b.size(), n));
```
Random tests should iterate to exercise more cases.

We need a fragmented manged_bytes so the test actually tests something.

bhalevy replied:
> Random tests should iterate to exercise more cases.

I wanted to ensure we test both small and large sizes, but doing a number of iterations is possible.

> We need a fragmented manged_bytes so the test actually tests something.

right, will change

denesb commented on utils/managed_bytes.hh:
```diff
@@ -394,6 +403,35 @@ public:
+    }
+
+    managed_bytes_view_fragment_iterator& operator++() noexcept {
+        remove_prefix(_current_fragment.size());
```
I think this is too heavy weight. This should be like 3 lines of code, no need to call the huge `remove_prefix()` to replace it.

bhalevy replied:
If it wasn't for `substr` I'd say you're right, but the way `substr` is defined, it may truncate the managed_bytes_view size so the iterator needs to take that into account and adjust the last fragment (dropping everything after it).

The reason `substr` is not actually truncating the view is that otherwise, it'll have to mess with `blob_storage`.
It can't modify the `blob_storage` it points to. They belong to the underlying `manage_bytes`.
So we'll need to provide it an interface to copy the last `blob_storage` fragment and modify its `frag_size` and `next` members.

denesb replied:
There is a simple solution to this: have a `_remains` member, from which you substract each time you move to the next fragment. When dereferencing you return `bytes_view(_current_fragment.data(), std::min(_remains, _current_fragment.size())`.

denesb replied:
In case the last fragment should be truncated (because of `substr()`) this will automatically return a `bytes_view()` with just the required part of it.

bhalevy replied:
Sounds good, I'll try that.

denesb commented on utils/managed_bytes.hh:
```diff
@@ -394,6 +403,35 @@ public:
+    friend class managed_bytes_view_fragment_iterator;
+};
+
+class managed_bytes_view_fragment_iterator : public managed_bytes_view_base {
```
There is no reason to inherit from the view, this is just confusing. I know you did it to reuse `remove_prefix()` but I think we shouldn't use that anyway.

denesb replied:
Also the required member types of an iterator are missing.

bhalevy replied:
which methods are missing?
There are: `operator*()`, `operator->()`, `operator++()`, `operator++(int)`
and they seem to be sufficient for all purposes in this series :)

denesb replied:
```c++
        using iterator_category	= std::forward_iterator_tag;
        using value_type = fragment_type;
        using pointer = const fragment_type*;
        using reference = const fragment_type&;
        using difference_type = std::ptrdiff_t;
```
These guys.

denesb replied:
At least make this inheritance private, code sharing is its only goal anyway.

bhalevy replied:
`manged_bytes_view` calls for `it.remove_prefix`, maybe we can friend it and hide it from external users.

denesb replied:
This violates the single responsibility principle. The iterator is now also tasked with mutating the underlying view, in fact it is also a view itself (partially), while it should only know how to iterate over the container.

denesb commented on utils/managed_bytes.cc:
```diff
@@ -23,8 +23,44 @@
+        advance_dst(n);
+        it.remove_prefix(n);
+    }
+    assert(!b);
```
Don't we already have this code in the `managed_bytes` copy constructor? We should reuse it.

bhalevy replied:
They are different, as the former copies `managed_bytes` and this one copies a `managed_bytes_view`.
Maybe if we change the former to use the `fragment_iterator` we could use common code.

denesb replied:
I think that would be best.

denesb commented on utils/managed_bytes.cc:
```diff
@@ -171,8 +178,15 @@ void managed_bytes_view_base::remove_prefix(size_t n) {
 
 managed_bytes_view
 managed_bytes_view::substr(size_t offset, size_t len) const {
```
We should have a special case for no `len` (take all). Maybe default `len` to `-1`.

bhalevy replied:
sounds good, maybe define managed_bytes_view::npos as -1.

denesb replied:
Yes.

denesb commented on utils/managed_bytes.hh:
```diff
@@ -462,6 +462,17 @@ public:
     size_t size_bytes() const noexcept { return size(); }
+
+    size_t hash() const noexcept {
+        if (!_next_fragments) {
```
`with_linearized()` does this also, so no need to do this here.

bhalevy replied:
right, thanks.

avikivity committed:
allocation_strategy: set preferred max contiguous allocation to 128k for standard allocations

Now that managed_bytes and its users do not assume that a managed_bytes
instance allocated using standard_allocation_strategy is non-fragmented,
we can set the preferred max contiguous allocation to 128k. This causes
managed_bytes to fragment instances that are larger than this size.

Note that managed_bytes is the only user.

bhalevy commented:
Latest version at 4dd4d9b:
* Still need to force fragmentation of `managed_bytes` but that can be done in a followup.
* And `fragment_iterator` is still based on `managed_bytes_view_base` which @denesb doesn't like, but it has exactly what the iterator needs.
* I defined an adapter class - `managed_bytes_fragment_range_view` conforming to `FragmentRange`, returned by `managed_bytes{,view}::as_fragment_range()`
* and implemented `managed_bytes_view::npos` arg to `substr` to return a suffix starting at `offset`
* also simplified fragment_iterator `operator++` to skip the current fragment more efficiently, by refactoring `remove_current` out of `remove_prefix`
* in addition, I implemented `managed_bytes_view::operator<=>` that calls `compare_unsigned`
* and cleaned up further the unit tests to test just one random case as the tests can run multiple iterations using an external options to `test.py`

avikivity commented:
@bhalevy latest commit on my branch will fragment m_b that are larger than 128k.

avikivity commented:
> Latest version at [4dd4d9b](https://github.com/denesb/scylla/commit/4dd4d9be21bb0c69b5a17d974ad22d4e49a123b2):
> * in addition, I implemented `managed_bytes_view::operator<=>` that calls `compare_unsigned`

That is risky, bytes::value_type is int8_t which compares differently.

Let's leave such enhancements for later.

denesb commented on utils/managed_bytes.hh:
```diff
@@ -414,12 +454,57 @@ public:
+    size_t size_bytes() const noexcept { return _view.size(); }
+    bool empty() const noexcept { return size_bytes() == 0; }
+private:
+    managed_bytes_view _view;
```
We usually place members on the top of the class definition.

denesb commented on utils/managed_bytes.hh:
```diff
@@ -414,12 +454,57 @@ public:
+    {}
+    explicit managed_bytes_fragment_range_view(managed_bytes_view&& mv) noexcept
+            : _view(std::move(mv))
+    {}
```
These can be `= default`.

denesb replied:
Oh, nevermind. :)

avikivity commented:
@bhalevy reminder that I need a guide about which patch goes where

denesb committed:
managed_bytes: managed_bytes_view_base: inline methods

The class will become a template.

denesb committed:
managed_bytes: managed_bytes_view: inline methods

The class is about to become a template.

denesb committed:
managed_bytes: make managed_bytes_view and friends templated on mutable_view

So we can have mutable and non-mutable views.

bhalevy committed:
test: managed_bytes_test: verify fragments

denesb committed:
managed_bytes: add bytes iterator

bhalevy commented:
Changes in this version:
* privately inherit `managed_bytes_view_fragment_iterator : managed_bytes_view_base`
* move `_view` member to top of class
* revert `operator<=>`
* fix: use `managed_bytes_view&` in [in]equality functions rather than `managed_bytes_view_base&`
* fix: fixed `appending_hash`
  - expose `update_appending_hash(Hasher& bytes_view)` and call from `appending_hash<managed_bytes_view>::operator()`
* rebase on top updated `avikivity/managed_bytes-no-data-accessor` with 128KB alloc limit
* test: add `max_size` variable
* test: add `verify_fragments` (for debugging, not sure that it's needed to be merged eventually)

denesb committed:
utils/fragment_range: add copy() algorithm

denesb committed:
managed_bytes: fixes

bhalevy committed:
managed_bytes, bytes_view: implement hash using appending_hash

Use appending_hash to implement std::hash<bytes_view> and
std::hash<managed_bytes_view> so that they would agree with
each other regardless of fragmentation, and with requiring
to serialize the managed_bytes_view.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy commented:
In  4385feb:
* added "managed_bytes, bytes_view: implement hash using appending_hash"
  - added `test_managed_bytes_view_hash`
* fixed `managed_bytes_view::substr`
* improved `managed_bytes_test.verify_fragments`

denesb committed:
managed_bytes_iterator: add operator+=() and operator+()

denesb committed:
managed_bytes: implement begin()

denesb committed:
managed_bytes: managed_bytes_basic_view: implement converting constructor
